source/common/upstream/cluster_manager_impl.cc

```diff
@@ -430,6 +432,24 @@ void ClusterManagerImpl::createOrUpdateThreadLocalCluster(ClusterData& cluster)
           });
 }
 
+void ClusterManagerImpl::createOrUpdateThreadLocalClusterInternal(
+    std::function<void()> thread_local_clustercb) {
+  if (tls_->isMainThread()) {
+    tls_->runOnAllThreads(thread_local_clustercb);
+  } else {
+    Thread::CondVar tls_update_;
+    Thread::MutexBasicLockable tls_lock_;
+    Thread::ReleasableLockGuard lock(tls_lock_);
+    main_thread_dispatcher_.post([this, &thread_local_clustercb, &tls_update_]() -> void {
```

Comment:
Here I tried to wait till all worker's are finished with another variant of runAllThreads but it does not work because this thread would have a lock and post would not be able to enter this. Any ideas on how we can wait on all worker threads to finish? But this seems to work fine, tested from Lua.

Reply:
Also any guidance on how I can test this in a unit test? How do I call this from a worker thread? Is there any sample that i can look at?

Reply:
I would like to step back and discuss the overall design for this feature. Here is how I would recommend we go about implementing it:

https://github.com/envoyproxy/envoy/blob/49b122d85aa8553798fc5817cdb764c217da2809/include/envoy/upstream/cluster_manager.h#L78-L79

At this API level, I would add a new function called `addOrUpdateClusterCrossThread`. This idea is that this function can only be called from worker threads. `addOrUpdateCluster` can only be called from the main thread. We should verify this using asserts.

Essentially, at an API level, we are going to make the caller deal with the fact that if they create a cluster off of the main thread they will need to deal with a continuation for when the cluster actually exists. Note, this is not as simple as passing a lambda that gets called (though this will be part of the call signature). The reason for this is that the callback must be cancelable. This is the reason it will be easier to have a new function, because instead of returning `void`, the cross thread version should return a handle to the pending cluster creation. This handle should have a `cancel()` function to cover the case where for example a filter tries to create a cluster but the request is reset/destroyed before cluster creation is complete.

For the actually cross-thread cluster creation, here is how it should work:

1. Register intent to create the cluster on the worker thread (save continuation, return a handle that can be canceled, etc.). You will need to think about corner cases such as the user creating a cluster that already exists. Maybe others.
2. Post a message to the main thread to create the cluster, passing all the data required to post back to the worker thread that initiated the call.
3. Create the cluster on the main thread (dealing with the fact that it might already exist).
4. Either the creation will post messages to all worker threads, or if it already exists, we must post a message back to the worker thread that initiated the call to unblock it.
5. On the worker thread, deal with unblocking any waiting creators (and of course if they canceled don't worry about it).

Hopefully that gives you enough to go on at a high level.

In terms of testing, I think probably testing this fully with unit tests will be a little difficult. There are two options:
1. Actually do some variant of cluster manager tests that use multiple threads.
2. Implement the Lua API as part of this change, which I think would not be that difficult (since Lua can just pass YAML, convert to proto, etc.), and then write Lua integration tests.

Reply:
@mattklein123 Thanks for the details. I have looked at it very briefly today and have some questions.
1. Is this how you are thinking that new method would look like?
ClusterFuturePtr ClusterManagerImpl::addOrUpdateClusterCrossThread(const envoy::api::v2::Cluster& cluster,
                                  const std::string& version_info,PostClusterCreationCb post_cluster_cb) {
  std::string cluster_name = cluster.name();
  std::async([this,&cluster,version_info,&post_cluster_cb]() { 
    main_thread_dispatcher_.post([ &cluster, version_info]()->void {
           // Write the logic of creating cluster here ..Specially 2,3,4 in your proposal.
    });
  }
 });
 return std::make_unique<ClusterFutureImpl>(cluster_name);
}
Are you thinking that we should run the main logic in std::async as shown above and return the handle?
ClusterFuture I am thinking is essentially the handle you are talking about. It will have a cancel method which caller would call. 
2. What should the cancel of handle expected to do? Should it remove the cluster?

Once I get clarity on this, I will start implementing it.

Reply:
No, not std::async. You should be using dispatcher post. You need to post to main thread, create the cluster, and then post back to all the workers. Each worker effectively needs to keep a list of pending cluster creation requests that will eventually be satisfied.

There are a few different ways of doing this, but basically you will create an object, insert it into a list, and return a handle which if destroyed/cancelled, removes the thing from the list. We have a few different examples of this in the code base. Have a look at: https://github.com/envoyproxy/envoy/blob/master/source/common/http/http1/conn_pool.h#L107, https://github.com/envoyproxy/envoy/blob/master/source/common/common/callback_impl.h, https://github.com/envoyproxy/envoy/blob/master/source/server/http/config_tracker_impl.h, etc.

Reply:
Ok. Thanks. Let me repeat what I understand so that there is no confusion.
1. When a worker thread calls the new API, it will create a call back object (that has cancel method), store the PostClusterCreationCb, insert it it in to a pending cluster creation list.
2. The PostClusterCreationCb holds the code to continue the request flow.
2. Post the message to main thread with the details of cluster creation - which would post back to worker threads. When the post back comes to the worker thread, we will create thread local cluster and see if there is a pending request (PostClusterCreationCb) and call it - so that ensures the request flow is continued after the cluster creation. So this way the worker thread is not blocked on this request - is that the intent?
3. When we return the Handle to the caller, in which case the continuation is triggered when the cluster creation on the worker is done - however if the request is reset, the caller would use the handle to cancel which would delete the object from the list so the continuation call back is never fired.
Is that correct understanding? Sorry - My knowledge on the lower level request flow in Envoy is pretty poor. That is why asking so many questions. Thanks much for your patience :-)

Reply:
Ok.cool. I am on PTO most of the week. I will make changes as soon as I am back.